src/plugins/matrixops/flip_operation.cpp

```diff
@@ -0,0 +1,219 @@
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+#include <hpx/util/optional.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <blaze/Math.h>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    match_pattern_type const flip_operation::match_data =
+    {
+        hpx::util::make_tuple("flip",
+        std::vector<std::string>{"flip(_1,_2)"},
+        &create_flip_operation, &create_primitive<flip_operation>,
+        "a, axis\n"
+        "Args:\n"
+        "\n"
+        "    a (vector or matrix) : a vector or a matrix\n"
+        "    axis (integer): an axis to flip along\n"
+        "\n"
+        "Returns:\n"
+        "\n"
+        "Reverses the order of elements in an array along the given axis")
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    flip_operation::flip_operation(primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    template <typename T>
+    primitive_argument_type flip_operation::flip1d(ir::node_data<T>&& arg,
+        val_type axis) const
+    {
+        if (axis != 0 && axis != -1)
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "flip_operation::flip1d",
+                generate_error_message(
+                    "the flip_operation primitive requires operand axis to be "
+                    "either 0 or -1 for vectors."));
+        }
+
+        auto v = arg.vector();
+        blaze::DynamicVector<T> result(v.size());
+        std::reverse_copy(std::begin(v), std::end(v), std::begin(result));
+
+        return primitive_argument_type{result};
+    }
+
+    template <typename T>
+    primitive_argument_type flip_operation::flip2d(ir::node_data<T>&& arg,
+        val_type axis) const
+    {
+        switch (axis)
+        {
+        case -2:
+            HPX_FALLTHROUGH;
+        case 0:
+            return flip2d_axis0(std::move(arg));
+
+        case -1:
+            HPX_FALLTHROUGH;
+        case 1:
+            return flip2d_axis1(std::move(arg));
+
+        default:
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "flip_operation::flip2d",
+                generate_error_message(
+                    "the flip_operation primitive requires operand axis "
+                    "to be between -2 and 1 for matrices."));
+        }
+    }
+
+    template <typename T>
+    primitive_argument_type flip_operation::flip2d_axis0(
+        ir::node_data<T>&& arg) const
+    {
+        using phylanx::util::matrix_row_iterator;
+
+        auto m = arg.matrix();
```

Comment:
Same here, if `arg` is not a reference the allocation of the result can be avoided.